chains/ethereum/listener.go

```diff
@@ -88,51 +152,43 @@ func buildQuery(contract ethcommon.Address, sig EventSig, startBlock *big.Int) e
 	return query
 }
 
+func buildSingleBlockQuery(contract ethcommon.Address, sig EventSig, startBlock *big.Int) eth.FilterQuery {
+	query := eth.FilterQuery{
+		FromBlock: startBlock,
+		ToBlock:   startBlock,
+		Addresses: []ethcommon.Address{contract},
+		Topics: [][]ethcommon.Hash{
+			{sig.GetTopic()},
+		},
+	}
+	return query
+}
+
 // RegisterEventHandler creates a subscription for the provided event on the bridge bridgeContract.
 // Handler will be called for every instance of event.
 func (l *Listener) RegisterEventHandler(subscription EventSig, handler evtHandlerFn) error {
-	evt := subscription
-	query := buildQuery(l.cfg.bridgeContract, evt, l.cfg.startBlock)
-	eventSubscription, err := l.conn.subscribeToEvent(query)
-	if err != nil {
-		return err
+
+	if l.subscriptions[subscription] != nil {
+		return fmt.Errorf("event %s already registered", subscription)
 	}
-	l.subscriptions[subscription] = eventSubscription
-	go l.watchEvent(eventSubscription, handler)
-	l.log.Debug("Registered event handler", "bridgeContract", l.cfg.bridgeContract, "sig", subscription)
+	sub := new(Subscription)
+	sub.handler = handler
+	sub.signature = subscription
+	l.subscriptions[subscription] = sub
+
 	return nil
-}
 
-// watchEvent will call the handler for every occurrence of the corresponding event. It should be run in a separate
-// goroutine to monitor the subscription channel.
-func (l *Listener) watchEvent(eventSubscription *ActiveSubscription, handler evtHandlerFn) {
-	for {
-		select {
-		case evt := <-eventSubscription.ch:
-			m := handler(evt)
-			err := l.router.Send(m)
-			if err != nil {
-				l.log.Error("subscription error: cannot send message", "sub", eventSubscription, "err", err)
-			}
-		case err := <-eventSubscription.sub.Err():
-			if err != nil {
-				l.log.Error("subscription error", "err", err)
-			}
-		}
-	}
 }
 
 // Unsubscribe cancels a subscription for the given event
 func (l *Listener) Unsubscribe(sig EventSig) {
-	if _, ok := l.subscriptions[sig]; ok {
-		l.subscriptions[sig].sub.Unsubscribe()
-	}
+	delete(l.subscriptions, sig)
 }
 
 // Stop cancels all subscriptions. Must be called before Connection.Stop().
 func (l *Listener) Stop() error {
-	for _, sub := range l.subscriptions {
-		sub.sub.Unsubscribe()
+	for sig := range l.subscriptions {
+		delete(l.subscriptions, sig)
```

Comment:
we don't need to do anything here anymore, it's not really necessary to empty the map, will get GC'd

Reply:
I think keep `Stop` but `Unsubscribe` can be removed